### ISSUE TITLE
When creating entity with keyValue options, Orion auto data type detection is not correct

### ISSUE THREAD

**Issue by AtosInnovationBordeaux** [telefonicaid/fiware-orion#3750]
When creating an entity you could do it like this : 
```bash
curl --location --request POST 'https://api.labatosbordeaux.eu/v2/entities/?options=keyValues' --header 'Content-Type: application/json' --data-raw '{
   "id":"vehicle:WasteManagement:1",
   "type":"WasteManagement",
   "vehicleType":"lorry"
}'
```

Or without the options=KeyValues parameter :
```bash
curl --location --request POST 'https://api.labatosbordeaux.eu/v2/entities/' --header 'Content-Type: application/json' --data-raw '{
   "id":"vehicle:WasteManagement:2",
   "type":"WasteManagement",
   "vehicleType":{
        "value": "lorry"
    }
}'
```

And in both case the entity is the same.
But I don't manage to create an entity with a options=keyValues parameter with an entity that has a 'location' attribute : when doing 

```bash
curl --location --request POST 'https://api.labatosbordeaux.eu/v2/entities/' --header 'Content-Type: application/json' --data-raw '{
   "id":"vehicle:WasteManagement:3",
   "type":"WasteManagement",
   "vehicleType":"lorry",
    "location": {
        "type": "Point",
        "coordinates": [
            -0.56832066,
            47.49576
        ]
    }
}'
```

In this case, the result from curl --location --request GET https://api.labatosbordeaux.eu/v2/entities/vehicle:WasteManagement:2'

Would be 
```bash
"location": {
    "type": "StructuredValue",
    "value": {
        "type": "Point",
        "coordinates": [
            -3.164485592,
            40.627851337
        ]
    },
    "metadata": {}
},
```

And the type of location is ```StructuredValue```

My expectation would be ```geo:point``` or other geo:XXX depending of the type of the GeoJSON which allow the entity to be geolocalized and filter by location.

**Comment by AtosInnovationBordeaux** [comment]
Looking at the code it looks like there should be some additional tests done arround 
https://github.com/telefonicaid/fiware-orion/blob/20b7ef386d4a5a3d54211aa8e264e970dd0a4db6/src/lib/jsonParseV2/parseContextAttribute.cpp#L251

And Orion should check if the inner type is a geo:XXX and what kind of geojson that is to populate the attribute with the correct type

**Comment by fgalan** [comment]
After having a look to this, I think is not actually an issue/bug, but a consequence of how Orion works. I think the answer would be better as SOF answer than in an issue comment (as the former is more focused in explaing functionalty and the later in issues/bugs).

Thus, could you create a question post in StackOverflow with this (it should be easy... basically a copy-paste ;), please? I'll answer the question at SOF

After that please add the link to the post as comment in this issue, then close the issue. Thanks!

**Comment by kzangeli** [comment]
I'd say it's a missing functionality and an issue with the label "backlog" would fit pretty well as a github issue.
But then again, that issue could be added after responding and explaining on SOF.

**Comment by fgalan** [comment]
> I'd say it's a missing functionality and an issue with the label "backlog" would fit pretty well as a github issue.
> But then again, that issue could be added after responding and explaining on SOF.

Ok, let's wait until the explanation at SOF. @AtosInnovationBordeaux by the moment ignore my previous comment about closing the issue and don't do that (just add the link to the SOF question).

**Comment by AtosInnovationBordeaux** [comment]
I would argue that it is an important missing feature because without that we can't really say that Orion allow the creation of entities through the keyValues format.

SOF link : https://stackoverflow.com/questions/65807353/when-creating-entity-with-keyvalue-options-orion-auto-data-type-detection-does

**Comment by fgalan** [comment]
Question answered at SOF.

With regards to 

> we can't really say that Orion allow the creation of entities through the keyValues format.

Note that the creation of entities **with specific attribute types** has not ever been the purpose (for a design/requirements point of view) of the keyValues format. As I tell in SOF answer, keyValues mode is a helper. Nothing else, nothing more.

Thus, under my understanding, nothing else needs to be done and this issue can be closed.

**Comment by AtosInnovationBordeaux** [comment]
Thanks for your anwsers Firmin.

We think that this feature, although an extension of the original specs, allows a better user and developper experience, particulary with clients that want fast PoCs and we have some that already went that route (using keyValues).
They use Orion in mobile app related use cases where they have to search for entities through geolocation filters but are blocked by this.

**Comment by kzangeli** [comment]
I'd be willing to fix this issue, as I believe it's something we should have implemented when we implemented the "auto-attr-type" for key-values, but simply didn't think about.

However I need the OK from Telefonica before I start.

**Comment by fgalan** [comment]
>  They use Orion in mobile app related use cases where they have to search for entities through geolocation filters but are blocked by this.

I don't agree the use case is blocked as long as the client publishing the information can use the normalized API.

May I have a look at the source code of the client (please link with the pointer to the part doing the creating/update request towards CB ) to evaluate how difficult would be to change from keyValues mode to normalized, please?

**Comment by AtosInnovationBordeaux** [comment]
It is not possible to give access to our client's code unfortunatly. 

The issue is that they already implemented and deployed their solution using keyValues and don't want to reimplement their Orion API and don't understand why this is an issue and feel like it should be working like this as well.

Do you think this feature would be a good addition to Orion ?

**Comment by fgalan** [comment]
> Do you think this feature would be a good addition to Orion ?

It depends on what we mean by "feature":

* If we mean to implement value introspection so Orion looking to the value payload can infer if the attribute is GeoJSON or not, I think is not a good addition. Detecting if a given JSON is GeoJSON is far from easy... just have a look to [the GeoJSON schemas](https://github.com/geojson/schema) that would need to be implemented in Orion. Even in the case we manage to implement it, adding a validation step to every single update in keyValues mode would involve an underised performance penalty.
* If we mean to implement a new option to be used in combination with keyValues in updates, e.g. `?options=keyValues,preserveType`, so the attribute type is not modified upon updating it, I think is a good addition. So first the entity is created using normalized mode setting the attribute types and after that updates in keyValues will use these attribute types (including the geo:xxx type for the attribute meaning location).

**Comment by AtosInnovationBordeaux** [comment]
For 1. I think it would be enough to do the check for attribute "location" only ? What do you think @kzangeli ? 
Since the Fiware community entities all use the location attribute for geolocation purpose. 

I also agree that 2 is a nice addition as well.

**Comment by jmcanterafonseca** [comment]
I think the intent of the keyValues format was to provide a simplified representation to be consumed by readers but not a "convenient" but in the end semantically incomplete format for writers. Furthermore, I would not allow entity creation with keyValues format. 

Linked Data best practices advocate to label data items with their type, see for instance https://json-ld.org/spec/latest/json-ld-api-best-practices/#typed-objects

**Comment by kzangeli** [comment]
First of all, does Orion modify a type on an update, if the type isn't given ?
I hope not - that would not be good at all.
I thought that this was all about "default" type when the type **isn't** given.

To answer @AtosInnovationBordeaux , my idea is to instead of just using "StructuredValue" as default type for a value that is a JSON Object, I'd look inside the payload body to see whether the JSON object has exactly two children, named "type" and "coordinates" (regardless of the name of the attribute) and in such case use a more precise default type (Geo, something).

**Comment by AtosInnovationBordeaux** [comment]
> First of all, does Orion modify a type on an update, if the type isn't given ?
> I hope not - that would not be good at all.
> I thought that this was all about "default" type when the type isn't given.

I guess it may depend with the way the update is done (update attribute only, update entity, update with options=upsert)

From my tests 

- For upsert update it does change the type from "geo:json" to "StructuredValue"
- For attribute update (PUT on /v2/entities/<ID>/attrs/location/value) it does not

**Comment by fgalan** [comment]
> First of all, does Orion modify a type on an update, if the type isn't given ?
> I hope not - that would not be good at all.
> I thought that this was all about "default" type when the type **isn't** given.

Not fully sure (it should be tested for the different kind of updating operations Orion support) but I think the way it works right now is as follows: when update request doesn't include attribute type in the JSON, the preconditions described in "Partial Representations" in the NGSIv2 spec apply so a type is chosen based on that, potentially changing the existing attribute type if that one differs.

Thus, in this sense, my idea of implementing a `preserveType` option would be an improvement in this line, to avoid changing attribute types if the user doesn't want it.

> 
> To answer @AtosInnovationBordeaux , my idea is to instead of just using "StructuredValue" as default type for a value that is a JSON Object, I'd look inside the payload body to see whether the JSON object has exactly two children, named "type" and "coordinates" (regardless of the name of the attribute) and in such case use a more precise default type (Geo, something).

In that case, something like this will end to an undesired behavior (in fact, in a 500 error due to GeoJSON violation at MongoDB):

```
POST /v2/entities/E/attrs?options=keyValues
...
"myAttr": {
  "value": {
    "type": "foo"
    "coordiantes": "bar"
  }
}
...
```

The full solution for this is to base the payload introspection on full GeoJSON validation and the cost of that is unarfordable.

**Comment by fgalan** [comment]
> The intent of the keyValues format was to provide a simplified representation to be consumed by readers but not a "convenient" but in the end semantically incomplete format for writers. Furthermore, I would not allow entity creation with keyValues format.

I agree.

The driving use case for keyValues is rendering information for readers. For creating/updating entities, although the NGSIv2 spec supports this mode, it is not a good idea and normalized mode should be used instead.

**Comment by kzangeli** [comment]
Yeah, I have to say I agree on most of what's said as well.
But ... keyValues for updates is a nice feature and if we can fix it, it would be good for our users.

One detail on
```
PUT on /v2/entities/<ID>/attrs/?options=keyValues
```

The nature of PUT is to overwrite/replace, so here overwriting the type is something to expect.
So, a URI param `preserveType'  isn't valid for PUT if we want to be RESTful (not that I'm an expert), right?
Had it been a PATCH instead ...

For POST and PATCH, preserveType (if the type is not given), should be the default behaviour (and perhaps the only behaviour).
For NGSI-LD, we've agreed upon forbidding modification of types - both for entities and attributes

**Comment by fgalan** [comment]
> One detail on
> 
> ```
> PUT on /v2/entities/<ID>/attrs/?options=keyValues
> ```
> 
> The nature of PUT is to overwrite/replace, so here overwriting the type is something to expect.
> So, a URI param `preserveType' isn't valid for PUT if we want to be RESTful (not that I'm an expert), right?
> Had it been a PATCH instead ...

Good point. I think you are correct. PUT on `/v2/entities/<ID>/attrs` should not allow `preserveType`.

> For POST and PATCH, preserveType (if the type is not given), should be the default behaviour (and perhaps the only behaviour).

The problem of doing preserveType the default behaviour is that it would break backward compatibility. Thus, an explicit `option=preserverType` should be used.
